### ISSUE TITLE
Refactor FederatedHPA controller

### ISSUE THREAD

**Issue by Poor12** [karmada-io/karmada#3690]
Currently, most of codes in federatedHPA controller is lifted from hpa-controller in kubernetes.
There are still a large number of functions in the code that have too many parameters and high complexity.
To facilitate maintenance, we need to refactor federatedHPA-controller.

**Comment by yuriy-yarosh** [comment]
It would be nice if FederatedHPA support Knative KPA, as well.
KPA uses built-in metrics provider, called Activator, to measure the RPS and Concurrency for the incoming requests, and to downscale all services to zero, during idle operation. Which is imposible with common HPA setups.

People are using custom HPA + Keda integrations as well, but there's no real world guarantee that all the clusters load would be balanced. 

I don't really get the point of FederatedHPA, when KPA & HPA + Keda approach had been widespreaded already.
But solving the cross-cluster load balancing issue with custom metrics, at CDN level, for instance, makes a lot more sense to me.

**Comment by jwcesign** [comment]
Hi @yuriy-yarosh Nice to see you!

>  but there's no real world guarantee that all the clusters load would be balanced.

1. The workloads/pods in different member clusters selected by the same HPA CR/resource share the load of the application equally. For example, 10 pods of the application are spread into two member clusters with distribution `cluster1: 3 pods, cluster2: 7 pods`, so the 3 pods in cluster1 take 3/10 of total requests and 7 pods in cluster2 take 7/10 of total requests. Scenarios that don't meet the restriction are not considered in this proposal.
2. If the system does not meet the first requirement, some clusters' services may become overloaded while others remain underloaded.      

About how to make the load balance, you can use a GLB and route the traffic based on the replicas of the multiple clusters.

> I don't really get the point of FederatedHPA, when KPA & HPA + Keda approach had been widespreaded already.  

Now, HPA only works in a single cluster, when there are multiple clusters, it can scale the workloads across clusters. So FederatedHPA aims to scale the workloads across multiple clusters.
So If you want to use HPA and scale the workloads in multiple clusters, you need it to configure it in multiple clusters, and the scaling is not flexible, for instance, by weight(cluster1:cluster2=1:2, so if 6 pods should be scaled up, 2 pods in cluster1, 4 pods in cluster2).

> It would be nice if FederatedHPA support Knative KPA

It's quite difficult to support KPA, because it's a project-specific scaling policy.
So the better way is: Install knative in member clusters, and propagate the ksvc to member clusters with Karmada, Then KPA will work in member clusters(but couldn't scaling across clusters).

**Comment by yuriy-yarosh** [comment]
@jwcesign thank you for an elaborate explanation.

> but there's no real world guarantee that all the clusters load would be balanced.

Usually, I prefer to skip LB's entirely and host multiple API gateways for every AZ, perform routing at the DNS or CDN level, to overcome throughput and performance limitations of the target LB implementation. Having too many entry points makes it really hard to guarantee even load balancing, and common LB setups became obsolete for anything highload (e.g. DPDK/IPDK related).

I'm not particularly happy about cluster per AZ situation, but multi AZ deployments are not well supported enough,  yet.

> you can use a GLB and route the traffic based on the replicas of the multiple cluster

My specific use case involves routing traffic at the CDN level (with Cloudflare Workers tbe), so I'm performing traffic splitting there, and sending 60% / 40% as per your example. I'm disregarding CPU/RAM based HPA scaling, because it doesn't correlate with the exact throughput and concurrency numbers, but I'm isolating user sessions that consume too much compute resources, for any reason.

>  Install knative in member clusters, and propagate the ksvc to member clusters with Karmada,

Yes, this makes total sense.
Is it possible to add Concurrency / RPS related scheduling, alongside common CPU/RAM ?
So, it would be possible to get a CPU/RAM per user session metric and reuse it in routing policies ? (to route resource demanding users elsewhere)

Right now I'm solving it with a custom ebpf pixie script that handles jwt (jwe) session cookies, to differentiate between user sessions

**Comment by jwcesign** [comment]
> Is it possible to add Concurrency / RPS related scheduling, alongside common CPU/RAM for FederatedHPA ?

From what I know, the RPS/Concurrency in Knative could be collected by Prometheus so that it could be used as custom metrics, and FederatedHPA could work with the custom metrics.
Also, with the metrics, you could split your traffic at the CDN level with dynamic configuration.

**Comment by yuriy-yarosh** [comment]
> could be collected by Prometheus

Yes [it's collected](https://github.com/knative/serving/blob/main/pkg/activator/handler/concurrency_reporter.go#L214) via [legacy OpenCensus](https://pkg.go.dev/knative.dev/pkg/metrics#section-readme) :\ We had to replace it with a ebpf metric.

> FederatedHPA could work with the custom metrics.

I could propagate ebpf metrics there, as well.

> you could split your traffic at the CDN level with dynamic configuration.

That's how things work right now.
